src/pcapi/core/offers/api.py

```diff
@@ -351,16 +372,31 @@ def upsert_stocks(
             edited_stocks.append(edited_stock)
             stocks.append(edited_stock)
         else:
+            valid_activation_codes_exist = validation.valid_activation_codes_exist(
+                stock_data=stock_data,
+                offer=offer,
+            )
+            quantity = len(stock_data.activation_codes) if valid_activation_codes_exist else stock_data.quantity  # type: ignore[arg-type]
+
             created_stock = _create_stock(
                 offer=offer,
                 price=stock_data.price,
-                quantity=stock_data.quantity,
+                quantity=quantity,
                 beginning=stock_data.beginning_datetime,
                 booking_limit_datetime=stock_data.booking_limit_datetime,
             )
+
+            if valid_activation_codes_exist:
+                activation_codes_created = _create_activation_codes(
+                    stock_data.activation_codes,  # type: ignore[arg-type]
+                    stock_data.activation_codes_expiration_datetime,
+                    created_stock,
+                )
+                activation_codes.extend(activation_codes_created)
```

Comment:
(détail)

Je ne suis pas sûr que la fonction apporte grand chose. Le code n'est pas beaucoup plus long si on l'_inline_, et on gagne une indirection, ce qui me rend (je trouve) le code un peu plus lisible :

```suggestion
                activation_codes.extend([
                    ActivationCode(
                        code=code,
                        expirationDate=stock_data.activation_codes_expiration_datetime,
                        stock=created_stock,
                    )
                    for code in stock_data.activation_codes
                ])
```